File: Excel/430 Challenge.R
library(tidyverse)
library(readxl)

shift_left <- function(mat, shift_size) {
  n_cols <- ncol(mat)
  tibble_mat <- as_tibble(mat)
  shifted_df <- tibble_mat %>%
    pmap_dfr(., ~{
      row_values <- c(...)
      shifted_values <- c(row_values[(shift_size + 1):length(row_values)], rep(NA, shift_size))
      return(as_tibble(t(shifted_values)))
    })
  as.matrix(shifted_df)
}


for (i in 26:1) {
  M <- matrix(NA, nrow = 1, ncol = 52)
  M[1, 1:i] <- 1:i
  M <- t(apply(M, 1, rev))
  M = shift_left(M, 53-2*i)
  M[!is.na(M)] <- LETTERS[M[!is.na(M)]]
  
  if (i == 26){
    M_final <- M
  } else {
    M_final <- rbind(M_final, M)
  }
}

mf_df = M_final %>% as.data.frame()

writexl::write_xlsx(mf_df, "430 Excel solution.xlsx")

File: Excel/400-499/430/430 Challenge.py
import pandas as pd
import numpy as np
import openpyxl

def shift_left(mat, shift_size):
    n_cols = mat.shape[1]
    df = pd.DataFrame(mat)
    shifted_df = df.apply(lambda row: pd.Series(np.roll(row, -shift_size)), axis=1)
    shifted_df = shifted_df.iloc[:, :n_cols]
    return shifted_df.values

# make dict of letters to numbers
letters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
letter_dict = {}
for i in range(26):
    letter_dict[letters[i]] = i+1

M_final = None

for i in range(26, 0, -1):
    M = np.full((1, 52), np.nan)
    M[0, :i] = np.arange(1, i+1)
    M = np.flip(M, axis=1)
    M = shift_left(M, 53-2*i)

    if i == 26:
        M_final = M
    else:
        M_final = np.vstack((M_final, M))


mf_df = pd.DataFrame(M_final)
mf_df = mf_df.applymap(lambda x: letters[int(x)-1] if not np.isnan(x) else '')

mf_df.to_excel("430 Excel solution PY.xlsx", index=False)